hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieMergedLogReader.java

```diff
@@ -0,0 +1,145 @@
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.hudi.common.table.log.HoodieMergedLogRecordScanner;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+/**
+ * Record Reader implementation to read avro data, to support inc queries.
+ */
+public class HoodieMergedLogReader extends AbstractRealtimeRecordReader
+    implements RecordReader<NullWritable, ArrayWritable> {
+  private static final Logger LOG = LogManager.getLogger(AbstractRealtimeRecordReader.class);
+  private final HoodieMergedLogRecordScanner logRecordScanner;
+  private final Iterator<HoodieRecord<? extends HoodieRecordPayload>> logRecordsKeyIterator;
+  private ArrayWritable valueObj;
+  private boolean supportTimeStamp = false;
+
+  private int end;
+  private int offset;
+
+  public HoodieMergedLogReader(RealtimeSplit split, JobConf job, HoodieMergedLogRecordScanner logRecordScanner) {
+    super(split, job);
+    this.logRecordScanner = logRecordScanner;
+    this.end = logRecordScanner.getRecords().size();
+    this.logRecordsKeyIterator = logRecordScanner.iterator();
+    this.valueObj = new ArrayWritable(Writable.class, new Writable[getHiveSchema().getFields().size()]);
+  }
+
+  private Option buildGenericRecordwithCustomPayload(HoodieRecord record) throws IOException {
+    if (usesCustomPayload) {
+      return record.getData().getInsertValue(getWriterSchema());
+    } else {
+      return record.getData().getInsertValue(getReaderSchema());
+    }
+  }
+
+  @Override
+  public boolean next(NullWritable key, ArrayWritable arrayWritable) throws IOException {
+    if (!logRecordsKeyIterator.hasNext()) {
+      return false;
+    }
+    Option<GenericRecord> rec;
+    HoodieRecord currentRecord = logRecordsKeyIterator.next();
+
+    rec = buildGenericRecordwithCustomPayload(currentRecord);
+    // try to skip delete record
+    while (!rec.isPresent() && logRecordsKeyIterator.hasNext()) {
+      offset++;
+      rec = buildGenericRecordwithCustomPayload(logRecordsKeyIterator.next());
+    }
+    if (!rec.isPresent()) {
+      return false;
+    }
+
+    GenericRecord recordToReturn = rec.get();
+    if (usesCustomPayload) {
+      // If using a custom payload, return only the projection fields. The readerSchema is a schema derived from
+      // the writerSchema with only the projection fields
+      recordToReturn = HoodieAvroUtils.rewriteRecord(rec.get(), getReaderSchema());
+    }
+    ArrayWritable curWritable = (ArrayWritable) HoodieRealtimeRecordReaderUtils.avroToArrayWritable(recordToReturn, getHiveSchema());
+
+    if (arrayWritable != curWritable) {
+      final Writable[] arrValue = arrayWritable.get();
+      final Writable[] arrCurrent = curWritable.get();
+      if (arrayWritable != null && arrValue.length == arrCurrent.length) {
+        System.arraycopy(arrCurrent, 0, arrValue, 0, arrCurrent.length);
```

Comment:
`arrayWritable != null` should never reach because `arrayWritable.get()` is invoked.

Reply:
yes， i will remove it。